aesd-char-driver/main.c

```diff
@@ -85,159 +96,340 @@ ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    loff_t newpos;
+    struct aesd_buffer_entry *entryPtr = NULL;
+    struct aesd_circular_buffer *devBuffPtr = &aesd_device.dev_cb_fifo;
+    size_t entryPtrInd;
+
+    uint64_t devBuffSize;
+
+    switch (whence)
+    {
+    case 0: /* SEEK_SET */
+        entryPtr = aesd_circular_buffer_find_entry_offset_for_fpos(
+            devBuffPtr, off, &entryPtrInd);
+        if (entryPtr)
+        {
+            newpos = off;
+        }
+        break;
+
+    case 1: /* SEEK_CUR */
+        entryPtr = aesd_circular_buffer_find_entry_offset_for_fpos(
+            devBuffPtr, filp->f_pos + off, &entryPtrInd);
+        if (entryPtr)
+        {
+            newpos = filp->f_pos + off;
+        }
+        break;
+
+    case 2: /* SEEK_END */
+        bufferSize(devBuffPtr, &devBuffSize);
+
+        entryPtr = aesd_circular_buffer_find_entry_offset_for_fpos(
+            devBuffPtr, devBuffSize + off, &entryPtrInd);
+
+        if (entryPtr)
+        {
+            newpos = devBuffSize + off;
+        }
+        break;
+
+    default: /* can't happen */
+        return -EINVAL;
+    }
+    if (!entryPtr)
+        return -EINVAL;
+    filp->f_pos = newpos;
+    return newpos;
+}
+
+long aesd_ioctl(struct file *filp, unsigned int cmd, long unsigned int seekParams_usr_addr)
+{
+    // int err = 0, tmp;
+    int retval = 0;
+    struct aesd_circular_buffer *devBuffPtr = &aesd_device.dev_cb_fifo;
+    loff_t newpos = 0;
+    int index, currOffset = 0;
+    struct aesd_seekto seekParams;
+    unsigned long retVal=0;
+
+    /*
+     * extract the type and number bitfields, and don't decode
+     * wrong cmds: return ENOTTY (inappropriate ioctl) before access_ok()
+     */
+    if (_IOC_TYPE(cmd) != AESD_IOC_MAGIC)
+        return -ENOTTY;
+    if (_IOC_NR(cmd) > AESDCHAR_IOC_MAXNR)
+        return -ENOTTY;
+
+    /*
+     * the direction is a bitmask, and VERIFY_WRITE catches R/W
+     * transfers. `Type' is user-oriented, while
+     * access_ok is kernel-oriented, so the concept of "read" and
+     * "write" is reversed
+     */
+    // if (_IOC_DIR(cmd) & _IOC_READ)
+    // 	err = !access_ok_wrapper(VERIFY_WRITE, (void __user *)seekParams, _IOC_SIZE(cmd));
+    // else if (_IOC_DIR(cmd) & _IOC_WRITE)
+    // 	err =  !access_ok_wrapper(VERIFY_READ, (void __user *)seekParams, _IOC_SIZE(cmd));
+    // if (err) return -EFAULT;
+
+    switch (cmd)
+    {
+
+    case AESDCHAR_IOCSEEKTO:
+
+        retVal = copy_from_user(&seekParams,
+                                (void *)seekParams_usr_addr,
+                                sizeof(struct aesd_seekto));
+
+        if(retVal)
+        {
+            return -EFAULT;
+        }
+
+        // seekParams = (struct aesd_seekto*)(seekParams_usr_addr);
+
+        PDEBUG("K-space ArgX: %u\r\nK-Space ArgY: %u",seekParams.write_cmd,seekParams.write_cmd_offset);
+
+        // Range-checking
+        if (seekParams.write_cmd > (bufferLength(devBuffPtr) - 1))
```

Comment:
@ECleverito same comments here, I think you need a locking strategy for `devBuffPtr`.